moveit_core/robot_model/include/moveit/robot_model/joint_model.h

```diff
@@ -130,8 +130,14 @@ class JointModel
   /** \brief The datatype for the joint bounds */
   using Bounds = std::vector<VariableBounds>;
 
-  /** \brief Construct a joint named \e name */
-  JointModel(const std::string& name);
+  /**
+   * @brief      Constructs a joint named \e name
+   *
+   * @param[in]  name                   The joint name
+   * @param[in]  index                  The index of the joint in the RobotModel
+   * @param[in]  first_variable_index   The index of the first variable in the RobotModel
+   */
+  JointModel(const std::string& name, size_t joint_index, size_t first_variable_index);
```

Comment:
By adding them to the constructer we make them immutable from the interface of JointModel.  I think that is a large improvement.  Long term I think it would be useful to make the whole structure immutable as it'll make it much easier to reason about and therefore possible to use on multiple threads safely.

Reply:
One thing we can do to improve the niceness of the API of MoveIt is to remove invalid states as this removes pre-conditions you have to check.  This makes it much harder to write performant code that can use this data on multiple threads.

Reply:
Removing the index from the joint model and instead just providing that information from the robot model might be better but would be a much larger change.

Reply:
> Removing the index from the joint model and instead just providing that information from the robot model might be better but would be a much larger change.

That's what I meant. I think it's the better way long term, but we can merge your changes first.